server/src/main/java/org/candlepin/controller/ContentAccessManager.java

```diff
@@ -265,122 +260,124 @@ public ContentAccessCertificate getCertificate(Consumer consumer)
+
+        existing.setCert(createX509Cert(consumer, owner, serial, keyPair, oneYearValidity));
+        existing.setContent(this.createPayloadAndSignature(owner, consumer));
+        ContentAccessCertificate savedCert = this.contentAccessCertificateCurator.create(existing);
+        consumer.setContentAccessCert(savedCert);
+        this.consumerCurator.merge(consumer);
+        return savedCert;
+    }
 
-            CertificateSerial serial = new CertificateSerial(endDate);
-            // We need the sequence generated id before we create the Certificate,
-            // otherwise we could have used cascading create
-            serialCurator.create(serial);
+    private ContentAccessCertificate updateScaCertificate(Consumer consumer, Owner owner,
+        ContentAccessCertificate existing) throws GeneralSecurityException, IOException {
+        Date now = new Date();
+        Date expiration = existing.getSerial().getExpiration();
+        boolean isX509CertExpired = expiration.before(now);
 
+        if (isX509CertExpired) {
+            Validity oneYearValidity = Validity.oneYear();
             KeyPair keyPair = this.keyPairCurator.getConsumerKeyPair(consumer);
-            byte[] pemEncodedKeyPair = pki.getPemEncoded(keyPair.getPrivate());
-
-            X509Certificate x509Cert = this.createX509Certificate(consumer, owner,
-                BigInteger.valueOf(serial.getId()), keyPair, startDate, endDate);
-
-            existing = new ContentAccessCertificate();
+            CertificateSerial serial = createSerial(oneYearValidity);
             existing.setSerial(serial);
-            existing.setKeyAsBytes(pemEncodedKeyPair);
-            existing.setConsumer(consumer);
-
-            pem.append(new String(this.pki.getPemEncoded(x509Cert)));
-            existing.setCert(pem.toString());
-            consumer.setContentAccessCert(existing);
-
-            existing = this.contentAccessCertificateCurator.create(existing);
-            consumer = this.consumerCurator.merge(consumer);
-        }
-        else {
-            pem.append(existing.getCert());
+            existing.setCert(createX509Cert(consumer, owner, serial, keyPair, oneYearValidity));
+            this.contentAccessCertificateCurator.saveOrUpdate(existing);
         }
 
-        // Generate the cert body/payload as necessary
-        if (oeca == null || regenerate) {
-            log.info("Generating new SCA payload for organization \"{}\"...", owner.getKey());
-
-            String contentJson = this.createPayloadAndSignature(owner, env);
-            oeca = new OwnerEnvContentAccess(owner, env, contentJson);
-            oeca = this.ownerEnvContentAccessCurator.saveOrUpdate(oeca);
+        Date contentUpdate = owner.getLastContentUpdate();
+        boolean shouldUpdateContent = !contentUpdate.before(existing.getUpdated());
+        if (shouldUpdateContent || isX509CertExpired) {
+            existing.setContent(this.createPayloadAndSignature(owner, consumer));
```

Comment:
@Januson @Ceiu A bit of a crazy idea I wanted to fly by you two here for the race condition:
What if, after we run `createPayloadAndSignature()` (which we assume to be an expensive operation), we do a refresh of the Owner object to see if the lastContentUpdate has changed during the time it was running? And if it did change, we scrap the newly generated payload and return the cert as is? This way we would reduce even further the chance for the race condition. The client might have stale data, but at least they won't miss the content entirely, since they'll be able to pick it up next time they check in?

So something like:
```
 if (shouldUpdateContent || isX509CertExpired) {
  String payload = this.createPayloadAndSignature(owner, consumer);

  this.ownerCurator.refresh(owner);
  // New content update sneaked in; return without changes
  if (contentUpdate.before(owner.getLastContentUpdate())) {
    return existing;
  }

  // no new content sneaked in, persist changes  
  existing.setContent(payload);
  this.contentAccessCertificateCurator.saveOrUpdate(existing);
 }
```
Do you think this will work? Any downsides?

Reply:
This could work, but it has the downside of being rather awkward and confusing to users who receive stale content for a blip, and then get corrected.

If we're going to check this post-generation (which is fair), we should just make this part a post-check loop that'll immediately do the regeneration again if we detect an update on the org. It's effectively the same behavior, except we save the user the confusion and re-fetching of the cert.

Reply:
Based on discussion in irc: this idea is good, but we'll need to think it through a bit more, with some design/diagrams.

Reply:
@Ceiu I had not the time to formulate all the possible proposals to enhance this; given that is holding back this PR from getting merged, and that this PR can stand by itself, are you ok with merging this as is, and treating the enhancement(s) as separate tasks?

Reply:
Sure